### ISSUE TITLE
how to resolve error message unable to perform ptrace pokedata

### ISSUE THREAD

**Issue by thitiwutnew** [frida/frida#2782]
Unable to get frontmost application on SM A536E: unable to perform ptrace pokedata: I/O error

**Comment by nankingrider** [comment]
same  here

**Comment by syail** [comment]
https://github.com/ViRb3/magisk-frida

Try this. I have same issues,
But using the above module solved the problem.

**Comment by kaftejiman** [comment]
Not sure if this is caused by the same root issue, but in my case this started to happen after installing the latest Google System update (version 351110190). The bootstrapper code was being written into a non writable region, therefore the ptrace pokedata call fails with I/O error. To solve this:

**Short answer (workaround):**
Libc (remote/local) comparison fails, the attached patch implements a refined comparison that only checks the inode, path and device major.

**Longer answer (take it with a grain of salt):**
At the start of the injection flow a comparison of the remote and local libc takes place [here](http://github.com/frida/frida-core/blob/main/src/linux/frida-helper-backend.vala#L1088)). Libc is delivered as part of the runtime APEX, the apexd can mount the apex packages more than once each time with a separate loop device, system_server and frida obtain their APEX mount via separate activation or bind‐mount operations, therefore each see a different loop device (with different minor numbers). As a result of this, the local/remote libc comparison based on the identity being the dev (major+minor) and the inode fails. Therefore, the injection flow proceeds with the code swap route [here](https://github.com/frida/frida-core/blob/main/src/linux/frida-helper-backend.vala#L1098) during which frida tries to load the bootstrapper into a RX mapped region. Prior to the Google System update version 351110190 this secondary route was working fine as frida matches the segment of the boot image which is loaded into a RWX, after the update however the runtime workings are changed, as a result this memory region no longer exists, now the enumeration process matches the `memfd:jit-zygote-cache` region which is not writable, as a result the ptrace call [here](https://github.com/frida/frida-core/blob/main/src/linux/frida-helper-backend.vala#L2297) fails with I/O.

**Debugged on:**
Pixel 6
Android 12
Frida: 16.6.3
Google System version 351110190

[0001-android-libc-comparison-fix.patch](https://github.com/user-attachments/files/18470751/0001-android-libc-comparison-fix.patch)

**Comment by vaibhav423** [comment]
@kaftejiman can us share the patched file.

**Comment by ttmx** [comment]
Using the latest version (16.7.13) and have the same issue. Android 8, old phone.
`Failed to spawn: unable to perform ptrace pokedata: I/O error`

**Comment by swarupsro** [comment]
Facing the same issues!

![Image](https://github.com/user-attachments/assets/9f56618b-3413-4f89-8fb6-f2422ac99a43)

**Comment by vaibhav423** [comment]
@kaftejiman I don't think so . The error seems to be occurring randomly (kind of) , it usually takes a reboot or even two sometime to fix . I would appreciate if u could help me or atleast suggest me from where I should start to find a fix for the error "unable to perform ptrace pokedata: I/O error"

**Comment by vaibhav423** [comment]
> Using the latest version (16.7.13) and have the same issue. Android 8, old phone. `Failed to spawn: unable to perform ptrace pokedata: I/O error`

any fix

**Comment by vaibhav423** [comment]
> Facing the same issues!
> 
> ![Image](https://github.com/user-attachments/assets/9f56618b-3413-4f89-8fb6-f2422ac99a43)

any fix

**Comment by kl1de** [comment]
> [@kaftejiman](https://github.com/kaftejiman) I don't think so . The error seems to be occurring randomly (kind of) , it usually takes a reboot or even two sometime to fix . I would appreciate if u could help me or atleast suggest me from where I should start to find a fix for the error "unable to perform ptrace pokedata: I/O error"

Hi, have you been able to fix this error?

**Comment by vaibhav423** [comment]
works now on lineageos

**Comment by RITESHGUPTA-7** [comment]
> Unable to get frontmost application on SM A536E: unable to perform ptrace pokedata: I/O error

We can uninstall the latest ART, to resolve the error.

1. Take shell access to android.
2. Switch to super user.
3. Uninstall using command `pm uninstall com.google.android.art`.

**Comment by sertois** [comment]
> > Unable to get frontmost application on SM A536E: unable to perform ptrace pokedata: I/O error
> 
> We can uninstall the latest ART, to resolve the error.
> 
> 1. Take shell access to android.
> 2. Switch to super user.
> 3. Uninstall using command `pm uninstall com.google.android.art`.

I had the same issue and this was the only thing that fixed it for me. Mind explaining how this works? Isn't this package critical for running apps, how does anything still run without it?

**Comment by ymuuuu** [comment]
fixed by disabling enforce denylist in magisk settings

**Comment by ZetaReserX** [comment]
I fixed it with magisk module which disables selinux every time android starts

**Comment by alexlauerman** [comment]
Rebooting eventually fixed this for me. I had also tried removing art, which may have helped but didn't immediately fix anything. I suggest rebooting the phone if you have this issue.

Edit: Removing ART caused issues on my phone. i was no longer able to run frida the next day, and the error appeared to be related to art missing.

**Comment by syahmizul** [comment]
Fixed this by removing Magisk modules that tries to hide root,adb and stuff like that.Remove Shamiko,LSPosed.I only have SELinux Permissive module.Maybe just try removing all Magisk modules and set SELinux manually.

**Comment by 2914644189** [comment]
unable to perform ptrace pokedata: I/O error，请问现在这个问题有确切的解决方法了吗？如果有请告知我一下，我将不胜感激！
